### PR TITLE
Remove inherited create() method + disallow copying of storages

### PR DESCRIPTION
**Pull request by rschu1ze**
The original motivation for this commit was that shared_ptr_helper used std::shared_ptr<>() which does two heap allocations instead of make_shared<>() which does a single allocation. Turned out that
1. the affected code (--> Storages/) is not on a hot path (rendering the performance argument moot ...)
2. yet copying Storage objects is potentially dangerous and was previously allowed.

Hence, this change

- removes shared_ptr_helper and as a result all inherited create() methods,

- instead, Storage objects are now created using make_shared<>() by the caller (for that to work, many constructors had to be made public), and

- all Storage classes were marked as noncopyable using boost::noncopyable.

In sum, we are (likely) not making things faster but the code becomes cleaner and harder to misuse.

This is a continuation of PR #36438.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by kitaisreal** [comment]
@rschu1ze not worth at all. Additional heap allocation for storage does not make any issues to performance at all. Changing 202 to fix that is not necessary.

**Comment by kitaisreal** [comment]
@rschu1ze I am not sure why copying was an issue before ? Concrete storages does not have any public constructors. IStorage has deleted copy constructor. Goal of previous design was to prevent clients to create storages using something other than `create` method.

**Comment by rschu1ze** [comment]
Agree about the performance argument but that was only the starting point and when I began this PR I wasn't really aware that the allocations are done only rarely. The actual improvement (IMHO) is that storage classes can no longer be copied which is dangerous in general (think of partially filled buffers, duplicated connections to the same remote database etc.). Copying is now disallowed (via boost::noncopyable).

**Comment by antonio2368** [comment]
> Goal of previous design was to prevent clients to create storages using something other than create method.

That was the solution, but what was the motivation for it, what problem did it try to solve?

You still expose all the constructors by making `shared_ptr_helper` friend of the class, the only difference is you cannot create it on the stack. I don't see a reason why you would do that, copying of the object is deleted as you said.
If you want to expose only specific constructor you make it public, and the rest private.

In short, I think it adds unnecessary complexity without any benefits to the API so I think this is a nice change to make the code a bit more clear. The changes are also already made, a lot of files are modified, but not that many lines.

**Comment by alexey-milovidov** [comment]
I like the changes - the code became more simple, and we got rid of shared_ptr_helper.

**Comment by rschu1ze** [comment]
I think everyone agrees that performance will not improve. But that was anyways only the initial motivation.

Now that @kitaisreal brought it up: IStorage indeed has a deleted copy ctor / assignment op, so one cannot copy at least any classes inheriting from IStorage. I missed that, and it admittedly takes away the main point of this PR.

What remains is @antonio2368 point that `shared_ptr_helper` as a friend gets access to all private class internals. The helper class is small though and only creates a shared_ptr so the risk is small. Also, one could previously not create storage objects (and objects of other classes inheriting from `shared_ptr_helper`) on the stack but this is more about code aesthetics than a performance problem as we are usually not on a hot path.

Of course, I don't insist on getting this change in. Just as a proposal (@kitaisreal ?): Since inheriting from boost::noncopyable is redundant for storage classes, I could remove that? We'd still get rid of shared_ptr_helper (i.e. less friends) and allow to create objects on the stack.

**Comment by kitaisreal** [comment]
@rschu1ze after discussion with Alexey we think that shared_ptr_helper can be convenient to use, but it does not improve anything we can delete it, for concrete storages we can remove inheritance from boost::noncopyable then merge.

File: src/Databases/MySQL/DatabaseMaterializedMySQL.cpp [hunk 1]
```diff
@@ -137,7 +137,7 @@ StoragePtr DatabaseMaterializedMySQL::tryGetTable(const String & name, ContextPt
     if (context_->isInternalQuery())
         return nested_storage;
     if (nested_storage)
-        return std::make_shared<StorageMaterializedMySQL>(std::move(nested_storage), this);
+        return StorageMaterializedMySQL::create(std::move(nested_storage), this);
```
Comment (antonio2368) [inline src/Databases/MySQL/DatabaseMaterializedMySQL.cpp:140]: this change is the opposite of intended

File: src/Databases/MySQL/DatabaseMaterializedTablesIterator.h [hunk 2]
```diff
@@ -23,7 +23,7 @@ class DatabaseMaterializedTablesIterator final : public IDatabaseTablesIterator
 
     const StoragePtr & table() const override
     {
-        StoragePtr storage = std::make_shared<StorageMaterializedMySQL>(nested_iterator->table(), database);
+        StoragePtr storage = StorageMaterializedMySQL::create(nested_iterator->table(), database);
```
Comment (antonio2368) [inline src/Databases/MySQL/DatabaseMaterializedTablesIterator.h:26]: this change is the opposite of intended

File: src/Interpreters/DatabaseCatalog.cpp [hunk 3]
```diff
@@ -250,7 +250,7 @@ DatabaseAndTable DatabaseCatalog::getTableImpl(
         /// It's definitely not the best place for this logic, but behaviour must be consistent with DatabaseMaterializedMySQL::tryGetTable(...)
         if (!context_->isInternalQuery() && db_and_table.first->getEngineName() == "MaterializedMySQL")
         {
-            db_and_table.second = std::make_shared<StorageMaterializedMySQL>(std::move(db_and_table.second), db_and_table.first.get());
+            db_and_table.second = StorageMaterializedMySQL::create(std::move(db_and_table.second), db_and_table.first.get());
```
Comment (antonio2368) [inline src/Interpreters/DatabaseCatalog.cpp:253]: revert

File: src/Storages/Kafka/StorageKafka.h [hunk 4]
```diff
@@ -117,7 +115,7 @@ class StorageKafka final : public shared_ptr_helper<StorageKafka>, public IStora
     std::list<std::shared_ptr<ThreadStatus>> thread_statuses;
 
     /// Handle error mode
-    HandleKafkaErrorMode handle_error_mode;
+    // HandleKafkaErrorMode handle_error_mode; // NOLINT -- actually not used, needs checing
```
Comment (antonio2368) [inline src/Storages/Kafka/StorageKafka.h:118]: It's in settings now and it wasn't deleted, you can delete it.

File: src/Storages/StorageMongoDB.h [hunk 5]
```diff
@@ -49,7 +45,7 @@ class StorageMongoDB final : public shared_ptr_helper<StorageMongoDB>, public IS
     void connectIfNotConnected();
 
     const std::string host;
-    const uint16_t port; /// NOLINT
+    // const uint16_t port; /// NOLINT -- actually not used, needs checking
```
Comment (antonio2368) [inline src/Storages/StorageMongoDB.h:48]: same as host and options, only used to construct uri

File: src/Databases/MySQL/DatabaseMaterializedMySQL.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Databases/MySQL/DatabaseMaterializedMySQL.cpp:140]: Yepp, thanks, will fix.

With `shared_ptr_helper`/`create()` gone, this does in theory not even build. My preliminary dev setup has the strange issue that it doesn't build MySQL support by default, so I didn't notice this problem when building locally. I currently think it has to do with the CMake cache which stores MYSQL as an option, will try with a flushed CMakeCache.

File: src/Databases/MySQL/DatabaseMaterializedTablesIterator.h (hunk 2 above)
Comment (rschu1ze) [reply src/Databases/MySQL/DatabaseMaterializedTablesIterator.h:26]: Done, tnx.

File: src/Interpreters/DatabaseCatalog.cpp (hunk 3 above)
Comment (rschu1ze) [reply src/Interpreters/DatabaseCatalog.cpp:253]: Done.

File: src/Storages/Kafka/StorageKafka.h (hunk 4 above)
Comment (rschu1ze) [reply src/Storages/Kafka/StorageKafka.h:118]: Done, tnx.

File: src/Storages/StorageMongoDB.h (hunk 5 above)
Comment (rschu1ze) [reply src/Storages/StorageMongoDB.h]: Done, tnx.

File: src/Storages/StorageMongoDB.cpp [hunk 6]
```diff
@@ -39,7 +39,7 @@ StorageMongoDB::StorageMongoDB(
     const String & comment)
     : IStorage(table_id_)
     , host(host_)
-    , port(port_)
+    // , port(port_)
```
Comment (antonio2368) [inline src/Storages/StorageMongoDB.cpp:42]: Remove this
you can also remove host and options fields

Comment (rschu1ze) [reply src/Storages/StorageMongoDB.cpp]: Alright, done.

**Comment by nikitamikhaylov** [comment]
I loved `shared_ptr_helper`.. RIP or maybe revert?

**Comment by nikitamikhaylov** [comment]
Also it now possible to create a storage on stack, because all constructors became public

**Comment by nikitamikhaylov** [comment]
And if you create a class that inherits from std::enable_shared_from_this on a stack - a disaster may happen
